### PR TITLE
add trait `GetStringProperties`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- add trait `GetStringProperties<>`
- add function `getStringProperties()`
- add object `StringProperty`

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
example:

```C++
namespace PMacc
{
namespace traits
{
    template< >
    struct StringProperties< ::picongpu::leheSolver::LeheSolver >
    {

        static StringProperty get()
        {
            PMACC_AUTO(
                prop,
                ::picongpu::leheSolver::LeheSolver::getStringProperties()
            );
            prop["name"].value = "LeheSolver";  // overwrite the name of the yee solver (inherit all other properties)
            return prop;
        }
    };
} // namespace traits
} // namespace PMacc
```

or

```C++
struct YeeSolver
{
// ...
    static PMacc::traits::StringProperty getStringProperties()
    {
        PMacc::traits::StringProperty prop("name","YeeSolver");
        prop["order"] = "2";
        return prop;
    }
// ...
};
```

File: src/libPMacc/include/traits/GetStringProperties.hpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            return T_Type::getStringProperties();
+        }
+    };
+
+
+    /** get the properties of an object
+     *
+     * The struct `StringProperties<>` needs to be specialized to change the result
+     * of this trait for a user defined type.
+     * If there is no user defined specialization avail this trait inherit from
```
Comment (ax3l) [inline src/libPMacc/include/traits/GetStringProperties.hpp]: available

Comment (ax3l) [inline src/libPMacc/include/traits/GetStringProperties.hpp]: inherit**s**

File: src/libPMacc/include/traits/GetStringProperties.hpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    struct GetStringProperties : public StringProperty
+    {
+        GetStringProperties() : StringProperty( StringProperties< T_Type >::get() )
+        {
+        }
+    };
+
+    /** get the properties of an object instance
+     *
+     * same as `GetStringProperties<>` but accept a instance instead a type
```
Comment (ax3l) [inline src/libPMacc/include/traits/GetStringProperties.hpp]: accept**s**
a**n** instance

File: src/libPMacc/include/traits/GetStringProperties.hpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+     *
+     * \param a instance which should be queried
```
Comment (ax3l) [inline src/libPMacc/include/traits/GetStringProperties.hpp]: a instance that shall be queried

File: src/libPMacc/include/traits/GetStringProperties.hpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+     * \return StringProperty of the given instance
+     */
+    template< typename T_Type >
+    StringProperty
+    getStringProperties( const T_Type& )
+    {
+        return GetStringProperties<T_Type>()();
+    };
+
+} //namespace traits
```
Comment (ax3l) [inline src/libPMacc/include/traits/GetStringProperties.hpp]: space weirdness

File: src/libPMacc/include/traits/GetStringProperties.hpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    /** get the properties of an object instance
+     *
+     * same as `GetStringProperties<>` but accept a instance instead a type
+     *
+     * \param a instance which should be queried
+     * \return StringProperty of the given instance
+     */
+    template< typename T_Type >
+    StringProperty
+    getStringProperties( const T_Type& )
```
Comment (ax3l) [inline src/libPMacc/include/traits/GetStringProperties.hpp]: const?

**Comment by ax3l** [comment]
works like a charm :)

File: src/libPMacc/include/traits/GetStringProperties.hpp (hunk 5 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/traits/GetStringProperties.hpp]: it is no member function therefore const is not allowed

Comment (ax3l) [inline src/libPMacc/include/traits/GetStringProperties.hpp]: ah did not look straight at it

**Pull request merged**